pkg/webhook/validation.go

```diff
@@ -102,8 +102,11 @@ func (v *validator) validate(ctx context.Context, obj runtime.Object) (admission
 			sourceCount++
 			unionCount++
 
-			if len(configMap.Name) == 0 {
-				el = append(el, field.Invalid(path.Child("name"), configMap.Name, "source configMap name must be defined"))
+			if len(configMap.Name) == 0 && configMap.Selector == nil {
+				el = append(el, field.Invalid(path.Child("name", "selector"), configMap.Name, "source configMap name or label selector are both empty; one must be set"))
+			}
+			if len(configMap.Name) > 0 && configMap.Selector != nil {
+				el = append(el, field.Invalid(path.Child("name", "selector"), configMap.Name, "source configMap name and label selector are both set; only one must be set"))
```

Comment:
```suggestion
				el = append(el, field.Invalid(path, configMap, "source configMap name and label selector are both set; only one must be set"))
```

I think you need to take a look at the error message here. Not sure what would be most user-friendly message. Maybe take a look at the `oneOf` validation? This validation is what we need here, and is supported på OpenAPI, but not yet by controller-runtime.

Reply:
When I add the `oneOf` validation manually in the CRD spec, this is the error message that is returned:

```bash
The Bundle "test" is invalid:
* <nil>: Invalid value: "": "spec.sources[1].secret" must validate one and only one schema (oneOf). Found none valid
* spec.sources[1].secret.name: Required value
* spec.sources[1].secret.selector: Required value
* <nil>: Invalid value: "": "spec.sources[2].configMap" must validate one and only one schema (oneOf). Found none valid
* spec.sources[2].configMap.name: Required value
* spec.sources[2].configMap.selector: Required value
```

Unless you says otherwise, I think trying to modify the CRD afterwards to inject the `oneOf` validation can be a fragile solution. So instead, since controller-runtime does not support `oneOf` OpenAPI validation out of the box, I tried to return an error message as close as possible to what the OpenAPI spec returns, with some minor deviations to be as user-friendly as possible.

It goes like that:

```bash
Error from server (Forbidden): error when creating "bundle.yaml": admission webhook "trust.cert-manager.io" denied the request: 
[
spec.sources.[2].configMap: Invalid value: "name: ' ', selector: nil": must validate one and only one schema (oneOf): [name, selector]. Found none valid,
spec.sources.[1].secret: Invalid value: "name: allo, selector: {}": must validate one and only one schema (oneOf): [name, selector]. Found none valid
]
```

1. Let's look at the first error, when both `name` and `selector` are not set:

```text
spec.sources.[2].configMap: Invalid value: "name: ' ', selector: nil": must validate one and only one schema (oneOf): [name, selector]. Found none valid
```

here, I simply hardcoded in the error message an empty string as the value of `name` and `nil` as the value of `selector`.

2. Let's look at the second error, when both `name` and `selector` are set:

```text
spec.sources.[1].secret: Invalid value: "name: allo, selector: {}": must validate one and only one schema (oneOf): [name, selector]. Found none valid
```

here, I parametrized the value of the `name` fields using the value provided by the user, but instead of parametrizing as well the value of `selector`, which turns out to be an ugly stringified version of a Golang map, I instead hardcoded `{}` to indicate that the value is not nil, it is set to some object.


Let me know what you think

Reply:
Thanks for the detailed work on this, @ocampeau! As long as the field reference is correct (example: `spec.sources.[1].secret`) , I wouldn't worry too much about the following text. I like your suggestion of dropping the default string formatting of a Golang struct.

Reply:
Cool! Thanks I am glad you like it.